Python: fix packaging

henryiii opened the pull request:
Using Python for the python portion is better, especially for editable installs. Still seems to be an issue in scikit-build-core with the path on the editable install for the .so file, though, so will look into it.

henryiii committed:
chore: use Python for package copy

Signed-off-by: Henry Schreiner <henryschreineriii@gmail.com>

henryiii commented:
This won't work in an editable install:

https://github.com/spglib/spglib/blob/c3af1dcf01a04964f153286d9fac7aa67a867184/python/spglib/spglib.py#L50

Because the `.so` file is not next to the python file. The Python file is in the original source location, and the `.so` file is in site-packages. I'll need to think about how this should be done. In what case does it not trigger the ImportError?

henryiii commented on python/MANIFEST.in:
FYI, MANIFEST.in is a setuptools thing. Scikit-build-core uses `sdist.include` and `sdist.exclude`.

LecrisUT replied:
Indeed, I have that in a clean-up pr

LecrisUT commented:
Thanks for the help on this

LecrisUT commented:
> This won't work in an editable install:
> 
> https://github.com/spglib/spglib/blob/c3af1dcf01a04964f153286d9fac7aa67a867184/python/spglib/spglib.py#L50
> 
> Because the `.so` file is not next to the python file. The Python file is in the original source location, and the `.so` file is in site-packages. I'll need to think about how this should be done. In what case does it not trigger the ImportError?

Actually I think we can do this with RPATH. The first part is so that it can use the system's `libsymspg.so`, but if it cannot find it, check for the bundled version. But we can hijack this by either adding an RPATH to `_spglib.so` pointing to the installed location of `libsymspg.so`, or changing the needed library to be absolute path. The RPATH method is usually the case when you first build it and then it gets stripped by `cmake --install`. Let's see how can we tell cmake to add an RPATH on install: [`CMAKE_INSTALL_RPATH`](https://cmake.org/cmake/help/latest/variable/CMAKE_INSTALL_RPATH.html).

How about exposing two `tool.scikit-build` options (with more sensible naming):
- `installed-library-path`: a list of paths relative to `CMAKE_INSTALL_PREFIX` pointing to where the library paths are installed (maybe this is exposed by `cmake-file-api` so that it can be deduced from the target and dependencies). Use this as `CMAKE_INSTALL_RPATH`. Default to `CMAKE_INSTALL_LIBDIR`
- `install-rpath`: whether or not to install rpath (or to override the rpath list) on non-editable installs. Always append rpath on editable installs. Default false

LecrisUT committed:
Fix python path

Editable installs overwrite the paths relative to the package names, and it fails when directory relative paths are used

Signed-off-by: Cristian Le <cristian.le@mpsd.mpg.de>

LecrisUT commented:
Ok I have added a quick fix for the python paths to make it work with editable installs. I have also tested the RPATH and it works fine. Just needs an implementation in scikit-build-core ;)

For reference here is what I did manually:
```console
$ pip install -e .
$ cd venv/lib/python3.11/site-packages/spglib
$ patchelf --add-rpath $(pwd) _spglib.cpython-311-x86_64-linux-gnu.so
$ readelf -d _spglib.cpython-311-x86_64-linux-gnu.so
$ cd ../../../../../
$ pytest
```

The `patchelf` command should be equivalent with `CMAKE_INSTALL_RPATH`.

PS: the 3-4 rebuilding issue is also fixed by this
